sdks/go/pkg/beam/x/hooks/grpc/grpc.go

```diff
@@ -0,0 +1,76 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime/harness"
+	"github.com/apache/beam/sdks/go/pkg/beam/util/grpcx"
+	"google.golang.org/grpc"
+)
+
+var grpcHookKey = harness.HookKey("grpc")
+
+// Hook allow a runner to tailor various aspects of gRPC
+// communication with the FnAPI harness. Each member of the struct
+// is optional; the default behavior will be used if a value is not
+// supplied.
+type Hook struct {
+	// Dialer allows the runner to customize the gRPC dialing behavior.
+	Dialer func(context.Context, string, time.Duration) (*grpc.ClientConn, error)
+	// TODO(wcn): expose other hooks here.
+}
+
+var grpcHookRegistry = make(map[string]Hook)
+
+// RegisterGrpcHook registers a GrpcHook for the
+// supplied identifier. It panics if the same identifier is
+// registered twice.
+func RegisterGrpcHook(name string, c Hook) {
+	if _, exists := grpcHookRegistry[name]; exists {
+		panic(fmt.Sprintf("RegisterGrpcHook: %s registered twice", name))
+	}
+	grpcHookRegistry[name] = c
+
+	harness.RegisterInitHook("grpc", func(_ context.Context) error {
+		if grpc := runtime.GlobalOptions.Get(grpcHookKey); grpc != "" {
```

Comment:
<!--new_thread; commit:d3b9b8992429b98ae5429208961299a90dcdec50; resolved:0-->
grpc is a package identifier in this file already. Could we rename this variable to something that doesn't conflict? Given the short scope (2 lines), and it's purpose as a key, perhaps k?